File: Day100/Part7/dynamic_median.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <unordered_map>

using namespace std;

// Problem 7: Dynamic Median
// Maintain a set of numbers with Add, Remove, and Query Median operations.
// 
// Algorithm: Two Heaps (Min-Heap and Max-Heap) with Lazy Deletion.
// - lower: Max-Heap (stores smaller half)
// - upper: Min-Heap (stores larger half)
// - Balance: lower.size() == upper.size() OR lower.size() == upper.size() + 1
// 
// Lazy Deletion:
// - Track count of values to be removed in an unordered_map.
// - unique clean() function that pops top of heaps if it's marked as deleted.

class DynamicMedian {
    priority_queue<long long> lower; // Max-heap
    priority_queue<long long, vector<long long>, greater<long long>> upper; // Min-heap
    unordered_map<long long, int> deleted;
    int lower_active = 0;
    int upper_active = 0;

    void clean_lower() {
        while (!lower.empty() && deleted[lower.top()] > 0) {
            deleted[lower.top()]--;
            lower.pop();
        }
    }

    void clean_upper() {
        while (!upper.empty() && deleted[upper.top()] > 0) {
            deleted[upper.top()]--;
            upper.pop();
        }
    }

    void balance() {
        // clean() before checking sizes? 
        // No, size() includes deleted elements. We need active counts.
        // But active counts are hard to track inside heap structure without cleaning.
        // Strategy: We maintain 'lower_active' and 'upper_active' counters manually.
        // We only clean() when accessing top().
        
        while (lower_active > upper_active + 1) {
            clean_lower();
            long long val = lower.top();
            lower.pop();
            lower_active--;
            
            upper.push(val);
            upper_active++;
        }
        
        while (upper_active > lower_active) {
            clean_upper();
            long long val = upper.top();
            upper.pop();
            upper_active--;
            
            lower.push(val);
            lower_active++;
        }
    }

public:
    void add(long long x) {
        // Heuristic insertion
        if (lower.empty()) {
            lower.push(x);
            lower_active++;
            return;
        }

        clean_lower();
        if (x <= lower.top()) {
            lower.push(x);
            lower_active++;
        } else {
            upper.push(x);
            upper_active++;
        }
        balance();
    }

    void remove(long long x) {
        // Check if exists?
        // Assuming valid removal query for this problem context.
        // Which heap is it in?
        // We don't know easily.
        // But we know x <= lower.top() -> likely in lower.
        // Wait, lazy deletion means we just Increment delete count.
        // But we must decrement active counter!
        // Where is it?
        // If x <= lower.top(), it SHOULD be in lower. 
        // Care needed: If x was in lower but now sitting in deleted set, active/non-active logic applies.
        
        // Correct Logic for Lazy Deletion Active Counts:
        // We need to know if 'x' is effectively in lower or upper.
        // We can compare with current EFFECTIVE split point.
        
        clean_lower();
        clean_upper();
        
        // If empty?
        if (lower_active == 0 && upper_active == 0) return;

        long long split = lower.top();
        
        if (x <= split) {
            if (lower_active > 0) { // Safety
                lower_active--;
                deleted[x]++;
            }
        } else {
            if (upper_active > 0) {
                upper_active--;
                deleted[x]++;
            }
        }
        balance();
    }

    double getMedian() {
        clean_lower();
        clean_upper();
        
        if (lower_active == 0) return 0;
        
        if (lower_active > upper_active) {
            return (double)lower.top();
        } else {
            return ((double)lower.top() + (double)upper.top()) / 2.0;
        }
    }
    
    // Debug helper
    int size() { return lower_active + upper_active; }
};

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    
    int Q;
    if (!(cin >> Q)) return 0;
    
    DynamicMedian dm;
    
    for(int i=0; i<Q; ++i) {
        string type;
        cin >> type;
        if (type == "add") {
            long long x;
            cin >> x;
            dm.add(x);
        } else if (type == "remove") {
            long long x;
            cin >> x;
            dm.remove(x);
        } else if (type == "median") {
            if (dm.size() == 0) cout << "Empty" << endl;
            else cout << dm.getMedian() << endl;
        }
    }
    
    return 0;
}

File: Day100/Part7/dynamic_median.py
import sys
import heapq

# Problem 7: Dynamic Median using Two Heaps (Lazy Removal)
# Python heapq is Min-Heap.
# Lower: Max-Heap (simulate with negated values).
# Upper: Min-Heap.

class DynamicMedian:
    def __init__(self):
        self.lower = [] # Max heap (negated)
        self.upper = [] # Min heap
        self.deleted_lower = {}
        self.deleted_upper = {}
        self.cnt_lower = 0
        self.cnt_upper = 0
        
    def _clean(self):
        while self.lower and self.deleted_lower.get(-self.lower[0], 0) > 0:
            val = -heapq.heappop(self.lower)
            self.deleted_lower[val] -= 1
            
        while self.upper and self.deleted_upper.get(self.upper[0], 0) > 0:
            val = heapq.heappop(self.upper)
            self.deleted_upper[val] -= 1
            
    def _balance(self):
        self._clean()
        
        while self.cnt_lower > self.cnt_upper + 1:
            val = -heapq.heappop(self.lower)
            self.cnt_lower -= 1
            heapq.heappush(self.upper, val)
            self.cnt_upper += 1
            self._clean()
            
        while self.cnt_upper > self.cnt_lower:
            val = heapq.heappop(self.upper)
            self.cnt_upper -= 1
            heapq.heappush(self.lower, -val)
            self.cnt_lower += 1
            self._clean()
            
    def add(self, x):
        self._clean()
        if not self.lower:
            heapq.heappush(self.lower, -x)
            self.cnt_lower += 1
        else:
            top_lower = -self.lower[0]
            if x <= top_lower:
                heapq.heappush(self.lower, -x)
                self.cnt_lower += 1
            else:
                heapq.heappush(self.upper, x)
                self.cnt_upper += 1
        
        self._balance()
        
    def remove(self, x):
        self._clean()
        if self.cnt_lower == 0 and self.cnt_upper == 0:
            return

        top_lower = -self.lower[0]
        
        if x <= top_lower:
            self.deleted_lower[x] = self.deleted_lower.get(x, 0) + 1
            self.cnt_lower -= 1
        else:
            self.deleted_upper[x] = self.deleted_upper.get(x, 0) + 1
            self.cnt_upper -= 1
            
        self._balance()
        
    def get_median(self):
        self._clean()
        if self.cnt_lower == 0:
            return 0
        
        if self.cnt_lower > self.cnt_upper:
            return float(-self.lower[0])
        else:
            return (-self.lower[0] + self.upper[0]) / 2.0

def solve():
    input_data = sys.stdin.read().split()
    if not input_data: return
    iterator = iter(input_data)
    
    try:
        Q = int(next(iterator))
        dm = DynamicMedian()
        results = []
        
        for _ in range(Q):
            type = next(iterator)
            if type == "add":
                x = int(next(iterator))
                dm.add(x)
            elif type == "remove":
                x = int(next(iterator))
                dm.remove(x)
            elif type == "median":
                if dm.cnt_lower + dm.cnt_upper == 0:
                    results.append("Empty")
                else:
                    med = dm.get_median()
                    # Format to match potential test cases (int if whole, else float)
                    if med.is_integer():
                        results.append(str(int(med)))
                    else:
                        results.append(str(med))
        
        sys.stdout.write('\n'.join(results) + '\n')
            
    except StopIteration:
        return

if __name__ == "__main__":
    solve()
